### ISSUE TITLE
Straight to the Gradient: Learning to Use Novel Tokens for Neural Text Generation

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#15]
## 논문 정보

- ICLR 2021 open review
- Anonymous authors

## 개요

- MLE로 학습한 생성 모델이 dull and repetitive phrases를 생성하는 문제를 해결하기 위한 방법론을 제안한 논문입니다.
- 이러한 text degeneration 문제를 해결하고자 제안된 대표적인 방법론은 크게 두 부류로 분류할 수 있습니다.
    1. **training-based methods**: negative tokens이 생성되지 않도록 하는 보조 로스를 추가하여 학습하는 unlikelihood training 등
    2. **decoding methods:** 문장 디코딩 단에서 randomness를 부여하여 토큰 반복 등을 막는 Nucleus/top-k sampling 등
- 논문에서 제안한 ScaleGrad는 학습 과정에서의 gradient를 인위적으로 조작하여 생성 모델이 novel tokens를 생성할 수 있도록 도와줍니다.
- 실험을 통해 스토리 생성 등의 open-ended generation 태스크뿐만 아니라 문서 요약, 기계 번역 등의 directed generation 태스크에서도 유효하다는 것을 확인했습니다.

## 방법론

- 모델이 ground truth에는 포함되어 있지 않은 novel tokens를 더 많이 생성할 수 있도록 유도하자.
- ground truth를 아래와 같이 표기할 때,
     <img width="239" alt="_2020-11-03__4 15 18" src="https://user-images.githubusercontent.com/49383323/98082812-b38d5100-1ebc-11eb-86a6-48c9319f8312.png">

    - novel token을 아래와 같이 정의합니다. 단 novel token의 집합은 각 타임 스텝 t에 의존하여 달라집니다.
        
        <img width="192" alt="_2020-11-03__4 18 35" src="https://user-images.githubusercontent.com/49383323/98082827-b8ea9b80-1ebc-11eb-9bd3-d45dcd5c6db1.png">

- MLE로 훈련되는 모델의 그래디언트를 분석해보자.
    - o_i^t를 i번째 인덱스의 토큰의 pre-softmax score라고 정의하고, p_k를 ground truth token에 대한 생성확률로 정의합니다. (k는 ground truth 토큰의 인덱스 번호)
    - logits에 대한 Loss의 편미분 값은 아래와 같이 나타낼 수 있습니다.
        
       <img width="645" alt="_2020-11-03__4 11 59" src="https://user-images.githubusercontent.com/49383323/98082862-c869e480-1ebc-11eb-91b5-5d3538600615.png">

        - ground truth token의 logit에 대한 gradient는 (p_k-1)
        - 그 이외의 다른 토큰들에 대해서는 p_i
    - 즉, 각 토큰에 대한 gradient는 생성확률 p에 직접적으로 연관되어 있기 때문에, 이 생성 확률 p를 조작함으로써 원하는 토큰들이 더 잘 생성될 수 있도록 모델을 학습시킬 수 있을 거라는 가설을 세울 수 있습니다.
- 아래와 같이 softmax output (tilde p 값)을 리노멀라이징합니다.
    - novel tokens에 속하는 토큰에 gamma값을 곱합니다 (gamma는 0~1).
    
      <img width="632" alt="_2020-11-03__4 29 03" src="https://user-images.githubusercontent.com/49383323/98082892-d91a5a80-1ebc-11eb-86b8-b7688c631c2e.png">

    - 노멀라이징 이후의 각 토큰 생성 확률은
        1. 토큰 i가 novel tokens 집합에 속하는 경우 (lambda는 0~1) 
            
      <img width="116" alt="_2020-11-03__4 40 14" src="https://user-images.githubusercontent.com/49383323/98082924-e9cad080-1ebc-11eb-9807-6f04bc253575.png">

        2. 토큰 i가 novel tokens 집합에 속하지 않는 경우 (alpha >1)
              
       <img width="106" alt="_2020-11-03__4 43 00" src="https://user-images.githubusercontent.com/49383323/98082936-ee8f8480-1ebc-11eb-8840-32c4512f5ba8.png">

- 노멀라이징 이후의 각 토큰에 대한 gradient는 다음과 같습니다.

     <img width="769" alt="_2020-11-03__4 36 02" src="https://user-images.githubusercontent.com/49383323/98083057-1979d880-1ebd-11eb-8b67-a87c018e4a31.png">

    - 즉, 모델의 학습은 grad norm이 0에 가까워지게 진행되기 때문에
    1. 어떤 토큰이 ground truth에 해당하는 토큰이고 novel tokens에 속한다면
        - p_tilde 값이 더 커지도록 모델의 학습이 진행됩니다.
    2.  어떤 토큰이 ground truth에 해당하는 토큰이고 novel tokens에 속하지 않으면
        - p_tilde 값이 더 작아지도록 모델의 학습이 진행됩니다.
    3. 어떤 토큰이 ground truth가 아닌 토큰이고, novel tokens에 속하지 않으면
        - p_tilde 값이 더 작아지도록 모델의 학습이 진행됩니다.
    4.  어떤 토큰이 ground truth가 아닌 토큰이고, novel tokens에 속한다면
        - p_tilde 값이 더 작아지도록 모델의 학습이 진행됩니다.

### Open-Ended Generation

- 언어 모델링, text auto-completion (50개의 토큰을 주고 나머지 100 토큰 맞히는 문제) 태스크에 대해 검증했습니다.
    - Wikitext-103 코퍼스를 학습 코퍼스로 이용합니다.
- 모델로는 pre-trained GPT2를 이용했습니다.
- 문장 디코딩 방식은 greedy 방식
- 평가 척도로는 PPL, Rep/l, uniq를 이용합니다.
    - Rep/l : 이전 l 토큰에 등장한 토큰이 반복되는 횟수
    - uniq: 생성된 unique한 토큰 개수
- 정량 평가 결과

<img width="952" alt="_2020-11-03__4 04 59" src="https://user-images.githubusercontent.com/49383323/98083117-31515c80-1ebd-11eb-8060-622fba7d8242.png">

- 정성 평가 결과

<img width="466" alt="_2020-11-03__4 05 42" src="https://user-images.githubusercontent.com/49383323/98083123-33b3b680-1ebd-11eb-8f99-52a7531c08cf.png">

### Directed Generation

- Image Paragraph Captioning
    - CNN Encoder+ LSTM Decoder
        
       <img width="474" alt="_2020-11-03__4 07 07" src="https://user-images.githubusercontent.com/49383323/98083137-39110100-1ebd-11eb-8d08-89823aec0057.png">

- Text Summarization
    - BertSum 모델
        
        <img width="548" alt="_2020-11-03__4 07 58" src="https://user-images.githubusercontent.com/49383323/98083153-3e6e4b80-1ebd-11eb-9152-35568f60bc29.png">
